Hides default fns inside Fuse impl to avoid exposing it to any crate

rakshith-ravi opened the pull request:
Fixes #70796 

@cuviper I've added some default, private traits to do the job for us. If required, I can expose them to a specific visibility if you want to call these functions for #70332

r? @cuviper

rakshith-ravi commented:
Oh also, I just submitted this to understand if my logic is correct. In case you think it's good, don't approve / merge yet. I'll add a few comments for others stumbling across this in the future and then it can be merged. Didn't want to do that yet in case my implementation logic was wrong

Dylan-DPC commented:
Added WIP to the title so it doesn't get merged, when you are ready you can remove it

rakshith-ravi commented:
Cool. Does the WIP mean it won't get reviewed? Cuz I want it to get reviewed before I can make further changes.

cuviper commented on src/libcore/iter/adapters/fuse.rs:
```diff
@@ -201,7 +172,131 @@ macro_rules! unchecked {
+impl<I> DoubleEndedIterator for Fuse<I> where I: DoubleEndedIterator + FusedIterator {}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<I> ExactSizeIterator for Fuse<I> where I: ExactSizeIterator + FusedIterator {}
```
Do we still need these impls? I think not, and I'm surprised this is accepted at all, but I guess they're acting as a sort of empty specialization here. The `stable` aspect should still be fine, since they're covered by the main impl.

rakshith-ravi replied:
Removed the empty impls

cuviper commented on src/libcore/iter/adapters/fuse.rs:
```diff
@@ -201,7 +172,131 @@ macro_rules! unchecked {
 }
```
This `unchecked!` macro is now defined far from its use, and those uses are also scattered more with how the `Impl` specializations are mixed. Let's move this up next to `fuse!` so at least the macros are organized together.

rakshith-ravi replied:
Done

cuviper commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 0a54a9403ddc41d13633c17a8d24c218dc563268 with merge 99ca7e89f078ee57814d4b9c5575f84f2a2d2329...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0a54a9403ddc41d13633c17a8d24c218dc563268","merge_sha":"99ca7e89f078ee57814d4b9c5575f84f2a2d2329"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=25852)
Build commit: 99ca7e89f078ee57814d4b9c5575f84f2a2d2329 (`99ca7e89f078ee57814d4b9c5575f84f2a2d2329`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=25852"},"merge_sha":"99ca7e89f078ee57814d4b9c5575f84f2a2d2329"} -->

rust-timer commented:
Queued 99ca7e89f078ee57814d4b9c5575f84f2a2d2329 with parent 93dc97a85381cc52eb872d27e50e4d518926a27c, future [comparison URL](https://perf.rust-lang.org/compare.html?start=93dc97a85381cc52eb872d27e50e4d518926a27c&end=99ca7e89f078ee57814d4b9c5575f84f2a2d2329).

rust-timer commented:
Finished benchmarking try commit 99ca7e89f078ee57814d4b9c5575f84f2a2d2329, [comparison URL](https://perf.rust-lang.org/compare.html?start=93dc97a85381cc52eb872d27e50e4d518926a27c&end=99ca7e89f078ee57814d4b9c5575f84f2a2d2329).

rakshith-ravi commented:
@cuviper a couple of questions:
- Now that we have removed those empty `impl`s, how do we know _for sure_ that the specialization is being called appropriately and not the default implementation? Is there any way to test that?
- Since we have an internal trait that hides the implementation of the specialization, will that be optimized away by the optimizer?
    - If it will be optimized away and removed after compile, how do we know that? Is there any way to see what gets optimized away and see what the final binary will look like (other than staring at bare-bone assembly, of course)?
    - If it's not gonna be optimized away, how much of a performance impact is it gonna be? Can we measure that somehow? Or at least estimate the additional CPU cycles it will take?

rakshith-ravi commented:
@cuviper sorry to bother you again, but did you get a chance to review this PR?

cuviper commented:
> * Now that we have removed those empty `impl`s, how do we know _for sure_ that the specialization is being called appropriately and not the default implementation? Is there any way to test that?

You could make an iterator that counts how many times `next()` is called, e.g.

```rust
struct Foo<'a>(&'a mut usize);
impl Iterator for Foo<'_> {
    type Item = ();
    fn next(&mut self) -> Option<()> {
        *self.0 += 1;
        None
    }
}

fn main() {
    let mut counter = 0;
    let mut foo = Foo(&mut counter).fuse();
    for _ in 0..10 {
        foo.next();
    }
    dbg!(counter);
}
```

This will print `1`, since the default `Fuse` intercepts further calls. But if we add:

```rust
impl std::iter::FusedIterator for Foo<'_> {}
```

Then it will print `10`, as the specialized `Fuse` forwards every call.

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0f5813bc3f0708ab50870eb3a263b1cc)

If you'd like to turn that into explicit tests, with and without `FusedIterator`, that would be great!

> * Since we have an internal trait that hides the implementation of the specialization, will that be optimized away by the optimizer?

It should be trivially inlined, yes, but that is more work for compilation time -- so I'm not surprised that the perf results show a little bit of a slowdown.

> If it will be optimized away and removed after compile, how do we know that? Is there any way to see what gets optimized away and see what the final binary will look like (other than staring at bare-bone assembly, of course)?

Well, before assembly there's also LLVM IR, so you could stare at that instead. :slightly_smiling_face: But without even bothering at the instruction level, you could just look at what functions are emitted. If there's no explicit function for the new impl layers, then those must have been inlined.

Beyond that, I think you'd have to have a concrete example to compare assembly before and after. The optimizer can do surprising things -- in the above playground, after optimization it's flattened entirely into `main`. It doesn't even iterate anything, just goes right into `Argument` setup for the `dbg!`, and this is *regardless* of whether it has `FusedIterator`! (I did confirm that the unoptimized LLVM IR for `Fuse::next` was indeed simpler with `FusedIterator`.)

cuviper commented on src/libcore/iter/adapters/fuse.rs:
```diff
@@ -173,35 +163,135 @@ impl<I> ExactSizeIterator for Fuse<I>
+// Fuse specialization trait
+// Iterators and DoubleEndedIterators cannot be overlapped successfully
+// So, they're separated into each it's own trait to provide internal implementations
+// Similarly, ExactSizeIterators cannot be overlapped, so requires its own trait
```
I think you _could_ bundle them all in one trait, if you add extra constraints on individual methods as needed. You can see an example of this in `ZipImpl`:
https://github.com/rust-lang/rust/blob/edc02580e4e80476ac1ded2cc1008eaf8b8400e6/src/libcore/iter/adapters/zip.rs#L83-L86

But I think it's also fine to have them in separate specialization traits.

rakshith-ravi replied:
Okay I think that might help the optimizer inline that entire thing away. I have no facts to back my claim, of course. Just going by intuition. I'll try that out and see how it performs.

cuviper replied:
I don't think it would matter for inlining -- the initial call depth is the same either way.

I think the code is fine your way, or we could consolidate like `ZipImpl` does. If nothing else, it would read a little easier, perhaps. You would have one `impl FuseImpl` with defaults, and just one specialized `impl FuseImpl` for `I: FusedIterator`. You don't need further specializations, because the extra double-ended/exact constraints are only needed on the specific methods.

```rust
// reduced example:
trait FuseImpl<I> {
    type Item;
    fn next(&mut self) -> Option<Self::Item>;
    fn next_back(&mut self) -> Option<Self::Item> where I: DoubleEndedIterator;
    fn len(&self) -> usize where I: ExactSizeIterator;
}
impl<I: Iterator> FuseImpl<I> for Fuse<I> {
    type Item = I::Item;
    default fn next(&mut self) -> Option<Self::Item> { ... }
    default fn next_back(&mut self) -> Option<Self::Item> where I: DoubleEndedIterator { ... }
    default fn len(&self) -> usize where I: ExactSizeIterator { ... }
}
impl<I: FusedIterator> FuseImpl<I> for Fuse<I> {
    fn next(&mut self) -> Option<Self::Item> { ... }
    fn next_back(&mut self) -> Option<Self::Item> where I: DoubleEndedIterator { ... }
    fn len(&self) -> usize where I: ExactSizeIterator { ... }
}
```

We should clarify this before merging though, because I don't think this "cannot be overlapped" comment is correct.

cuviper commented:
@Centril, since it was your request to internalize this, how do you feel about the slightly worse perf results?

Centril commented:
@cuviper Hmm; do you know where the regressions are coming from (I haven't checked the diff)? They don't look egregious, so imo correctness comes first in this case, but I wonder if we could have our cake and eat it too?

cuviper commented:
Nothing stands out to me -- but just intuitively, there's an extra layer of function calls that have to be inlined away. Even in debug mode, it's simply a little bit more code to compile.

> I wonder if we could have our cake and eat it too?

Well, we had chatted about a new language feature like `default(crate)` -- that would avoid needing nested calls to hide the specialization impls.

Centril commented:
Ah OK. :) `default(crate)` is probably a long way from being implemented.

My personal view is that, for now, we should accept the slight regression here, merge the PR, and try to recover it when we can do so in a correct way.

rakshith-ravi commented:
@bors try @rust-timer queue

bors commented:
@rakshith-ravi: :key: Insufficient privileges: not in try users

rust-timer commented:
Insufficient permissions to issue commands to rust-timer.

rakshith-ravi commented:
Gah!

cuviper commented on src/libcore/iter/adapters/fuse.rs:
```diff
@@ -44,13 +44,213 @@ macro_rules! fuse {
+}
+
+// Fuse specialization trait
+#[doc(hidden)]
```
The `#[doc(hidden)]` seems weird, but `ZipImpl` has this too. :man_shrugging:

rakshith-ravi replied:
Yeah, I'm guessing it's to keep plugins in VSCode, etc from randomly showing this up on auto-complete.

cuviper commented:
I don't expect performance will be much different than before, but we can try it.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit abe5973b9d1202f1a6460c0116f22fc33c9506c5 with merge 26002ac6b8bee06e3c810c453f5771d8b523b1da...
<!-- homu: {"type":"TryBuildStarted","head_sha":"abe5973b9d1202f1a6460c0116f22fc33c9506c5","merge_sha":"26002ac6b8bee06e3c810c453f5771d8b523b1da"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=26434)
Build commit: 26002ac6b8bee06e3c810c453f5771d8b523b1da (`26002ac6b8bee06e3c810c453f5771d8b523b1da`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=26434"},"merge_sha":"26002ac6b8bee06e3c810c453f5771d8b523b1da"} -->

rust-timer commented:
Queued 26002ac6b8bee06e3c810c453f5771d8b523b1da with parent 835428c35d785733e72bfbf32fc2f8fff3e50e63, future [comparison URL](https://perf.rust-lang.org/compare.html?start=835428c35d785733e72bfbf32fc2f8fff3e50e63&end=26002ac6b8bee06e3c810c453f5771d8b523b1da).

Mark-Simulacrum commented:
FYI: We just deployed [a PR](https://github.com/rust-lang/rustc-perf/pull/645) to the perf collector that in theory reduces noise, but may also introduce other unexpected effects. If the results are odd or otherwise unexpected, it may be worth re-running the try build and the perf run. (Indeed, I would likely recommend just doing so regardless, but up to you).

rust-timer commented:
Finished benchmarking try commit 26002ac6b8bee06e3c810c453f5771d8b523b1da, [comparison URL](https://perf.rust-lang.org/compare.html?start=835428c35d785733e72bfbf32fc2f8fff3e50e63&end=26002ac6b8bee06e3c810c453f5771d8b523b1da).

rakshith-ravi commented:
Wait, the CPU cycles has reduced by 10%?!? And the number of instructions have reduced by 20%?!?

Is this because of the new PR @Mark-Simulacrum?

I mean, I had a hunch that the perf would increase with my new changes, but not this much xD
10% seems like a lot.

Mark-Simulacrum commented:
We've seen improvements elsewhere, so it's most likely spurious, but let's try it again: @bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit abe5973b9d1202f1a6460c0116f22fc33c9506c5 with merge 6524bc32a6a2b550d6fbbb229227b9f5a3cfd876...
<!-- homu: {"type":"TryBuildStarted","head_sha":"abe5973b9d1202f1a6460c0116f22fc33c9506c5","merge_sha":"6524bc32a6a2b550d6fbbb229227b9f5a3cfd876"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=26484)
Build commit: 6524bc32a6a2b550d6fbbb229227b9f5a3cfd876 (`6524bc32a6a2b550d6fbbb229227b9f5a3cfd876`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=26484"},"merge_sha":"6524bc32a6a2b550d6fbbb229227b9f5a3cfd876"} -->

rust-timer commented:
Queued 6524bc32a6a2b550d6fbbb229227b9f5a3cfd876 with parent 534a41a32952d36ec73656357777ebbea707aeb4, future [comparison URL](https://perf.rust-lang.org/compare.html?start=534a41a32952d36ec73656357777ebbea707aeb4&end=6524bc32a6a2b550d6fbbb229227b9f5a3cfd876).

cuviper commented:
The new performance results look pretty tightly neutral -- which is a good thing!

@bors r+ rollup

bors commented:
:pushpin: Commit abe5973b9d1202f1a6460c0116f22fc33c9506c5 has been approved by `cuviper`

<!-- @bors r=cuviper abe5973b9d1202f1a6460c0116f22fc33c9506c5 -->
<!-- homu: {"type":"Approved","sha":"abe5973b9d1202f1a6460c0116f22fc33c9506c5","approver":"cuviper"} -->
